[Tracer][ONNX] Enable numel tracing

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74082
* __->__ #74081

Fixes #74047

Trace `tensor.numel()` as `aten::numel`, instead of a constant value. The logic is similar to tracing `tensor.size(dim)`. 
Replaces #74059

Differential Revision: [D35299848](https://our.internmc.facebook.com/intern/diff/D35299848)

BowenBao committed:
Enable numel tracing

clang-format

resolve onnx test failure

update expect file

[ghstack-poisoned]

BowenBao commented on test/onnx/expect/TestOperators.test_view_flatten.expect:
```diff
@@ -39,16 +121,16 @@ graph {
           }
           dim {
-            dim_value: 24
+            dim_param: "Reshape11_dim_1"
```
Note this dynamic change is introduced by `tensor.numel` being traced dynamically, and will be reverted in the next PR in the stack where constant folding for `onnx::ReduceProd` is introduced.

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6474,6 +6474,9 @@ def test_embedding_bag(self):
 
+    @disableScriptTest()  # error in propagate as assign input shape
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_embedding_bag_with_offset(self):
```
Similar to above, offset becomes dynamic, hence requires opset 11.
Next PR in the stack reverts this effect.

BowenBao committed:
Update on "[Tracer][ONNX] Enable numel tracing"


Fixes #74047

Trace `tensor.numel()` as `aten::numel`, instead of a constant value. The logic is similar to tracing `tensor.size(dim)`. 
Replaces #74059

[ghstack-poisoned]

BowenBao commented:
@albanD please let me know if this looks good to you.

albanD commented:
@BowenBao there are indeed precedents of special casing for tracing in these functions. So I guess this is ok to do.

BowenBao commented:
Thanks @albanD .

@malfet please import. Only the first PR needs importing as it modifies tracer, the second one can be auto merged.

BowenBao commented:
@albanD could you approve on behalf of the JIT changes?

@shubhambhokare1, @garymm please review ONNX side changes, thanks.

garymm commented on test/jit/test_tracer.py:
```diff
@@ -377,6 +377,18 @@ def test_trace_size(self):
+        x = torch.randn(2, 3, 4)
+        y = torch.randn(4, 5, 6)
+
+        # Check that it behaves as expected
```
Nit: this comment is pretty useless. All tests do that :-)

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6593,24 +6601,28 @@ def forward(self, input, batch1, batch2, alpha, beta):
-        self.run_test(model, (x,))
+        self.run_test(model, (x,),
+                      input_names=['x'], dynamic_axes={'x': [0, 1, 2]},
+                      test_with_inputs=[(x2, )])
```
```suggestion
                      test_with_inputs=[(x2,)])
```

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6593,24 +6601,28 @@ def forward(self, input, batch1, batch2, alpha, beta):
-        self.run_test(model, (x,))
+        self.run_test(model, (x,),
+                      input_names=['x'], dynamic_axes={'x': [0]},
+                      test_with_inputs=[(x2, )])
```
```suggestion
                      test_with_inputs=[(x2,)])
```

garymm reviewed (approved):
🦈

BowenBao committed:
Update on "[Tracer][ONNX] Enable numel tracing"


Fixes #74047

Trace `tensor.numel()` as `aten::numel`, instead of a constant value. The logic is similar to tracing `tensor.size(dim)`. 
Replaces #74059

[ghstack-poisoned]

albanD commented:
I'm afraid I don't know the tracer code at all, @eellison can you take a quick look to make sure it is ok please?

eellison commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -894,6 +894,27 @@ autograd::Variable getSizeOf(const autograd::Variable& var, int64_t dim) {
   return size_var;
 }
 
+autograd::Variable getNumelOf(const autograd::Variable& var) {
```
Could we refactor this with `getSizeOf` maybe ?

BowenBao replied:
I feel it would add more complexity for handling different arguments based on operator types, while currently we only need inserting a constant in graph for `dim` of `Size`. 
If we choose to address arguments outside of the function, then we'd need the same boilerplate for getting graph from tracing state. 
Please let me know if you have a better idea.

eellison replied:
yea it's fine as is

eellison reviewed (approved):
LGTM

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35299848).

BowenBao committed:
Update on "[Tracer][ONNX] Enable numel tracing"


Fixes #74047

Trace `tensor.numel()` as `aten::numel`, instead of a constant value. The logic is similar to tracing `tensor.size(dim)`. 
Replaces #74059

Differential Revision: [D35299848](https://our.internmc.facebook.com/intern/diff/D35299848)

[ghstack-poisoned]

BowenBao commented:
@malfet rebased

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35299848).
